### ISSUE TITLE
How to turn 1h to 4h candles?

### ISSUE THREAD

**Issue by Kalelv45** [ccxt/ccxt#6042]
MUST READ THIS BEFORE SUBMITTING ISSUES (read the link, then delete this message before submitting):

https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#how-to-submit-an-issue

Please make sure your local version of ccxt is up to date with this repo. You can check by comparing the output of `ccxt.version` to https://github.com/ccxt/ccxt/blob/master/package.json#L3

- OS: Mac OS
- Programming Language version: Python 3.7
- CCXT version: (last version)
- Exchange: Bitmex
- Method: fetch OLHCV

There is a way to change the 1h time frame to the 4h time frame, using the exact candles that bitmex uses? 
Does anyone have snippet for that???

**Comment by frosty00** [comment]
```
bitmex = ccxt.bitmex()
print(bitmex.fetch_ohlcv('ETH/USD', '1h')[::4])
```

you could also have more code to check that no candles are missing (ccxt does not check for that and assumes that the exchange returned the correct candles)

Let us know if you need anymore help )

**Comment by kroitor** [comment]
@frosty00 I think the question is how to aggregate 1h candles for BitMEX (that has emulated OHLCV and therefore can even serve incomplete candles) into 4h candles. However, `[::4]` will just take each 4th element... no?

The solution is probably to aggregate 1m candles to 4h in a loop to avoid gaps in the data.

**Comment by frosty00** [comment]
yeah you are right it will just take the fourth 1h candle...

**Comment by kroitor** [comment]
> yeah you are right it will just take the fourth 1h candle...

Yep, basically, we need a file in `examples/...` that would do such aggregation for the emulated OHLCV or true-OHLCV, however, i didn't have the time to write one yet...

**Comment by frosty00** [comment]
wb this

```
bitmex = ccxt.bitmex()
ohlcv = bitmex.fetch_ohlcv('ETH/USD', '1h')
aggregate = [[sum(hour[field_index] for hour in four_hours) / 4 for field_index in range(4)] for four_hours in zip(ohlcv[::4], ohlcv[1::4], ohlcv[2::4], ohlcv[3::4])]
pprint(aggregate)
```

also i had very little sleep last night but it may work

**Comment by kroitor** [comment]
> wb this

That won't work robustly either, since with emulated OHLCV we might get incomplete candles, therefore we need to aggregate 1m into 4h, not 1h into 4h. Volumes should be added, highs and lows should be min/maxed. Therefore it will become unreadable if we do it in a single line ) Nw, we will write a clean example.

**Comment by Kalelv45** [comment]
> @frosty00 I think the question is how to aggregate 1h candles for BitMEX (that has emulated OHLCV and therefore can even serve incomplete candles) into 4h candles. However, `[::4]` will just take each 4th element... no?
> 
> The solution is probably to aggregate **1m** (not 1h) candles to 4h in a loop – using a lower source timeframe like 1m to avoid gaps in the data.

I don't think you need the 1m candles, I think 1 hour would be just fine. 
You just need to index them well so it actually aggregate every 4 candle at the same time that bitmex use for the 4h candles. 
You have posted some where the option to aggregate 5 min to 15 min, I think 1h to 4h, would need just an extra parameter that index them the same way.

**Comment by hanyoonLDA** [comment]
Time will need to take the value from the first candle.
O will need to take the value from the first candle of the aggregate.
H will need to take the max value of the aggregated candles. 
L will need to take the low value of the aggregated candles.
C will need to take the value of the last candle of the aggregate.
V will need to be the sum of all candles in the aggregate.

**Comment by hanyoonLDA** [comment]
1m to 15min example using pandas. 
```
import pandas as pd

# Convert OHCLV values into Panda's DataFrame object. Set limit and since accordingly.
values = exchange.fetchOHCLV(marketname, '1m')

# Probably need to convert `values` into a dict before this step. You can probably figure it out.
df = pd.DataFrame(values)

# Dictionary to tell Pandas how to re-sample
ohlc_dict = {'Open':'first', 'High':'max', 'Low':'min', 'Close': 'last', 'Volume':'sum'}

# Resampling to 15m. 
df.resample('15M').apply(ohlc_dict)
```
Disclaimer: I have not tried it out myself, use it as a starting point and post your code & results ;)

**Comment by 4ext** [comment]
Thanks for the great discussion, I also need to do this... 
Does anyone have a suggestion as to how to find the correct interval start per exchange?
Fetch a daily candle to get the timestamp where it starts and align to this per interval? But what if there's no daily :/
Is there an easier way?

**Comment by kroitor** [comment]
@4ext 

> Does anyone have a suggestion as to how to find the correct interval start per exchange and timeframe?
> Fetch a daily candle to get the timestamp where it starts and use this to align intervals? 
> Is there an easier way?

This is possible, however, there's no easy single way to do this. You have to improvise, since it is exchange-specific, and not unified yet. So, your suggested solution would work for some exchanges.

Alternatively, you can:
- try fetching from the epoch date 1 Jan 1970, then see which candle is the first in the response, however, that might not really work for all exchanges.
- try getting the historical public trades for a market to determine the start of the trading (and, effectively, the start of OHLCV history).

> But what if there's no daily :/

Most exchanges do have it.

**Comment by 4ext** [comment]
Whoa! Thanks a lot for the very quick reply @kroitor !

> This is possible, however, there's no easy single way to do this. You have to improvise, since it is exchange-specific, and not unified yet. So, your suggested solution would work for some exchanges.
> 
> Alternatively, you can:
> 
>     try fetching from the epoch date 1 Jan 1970, then see which candle is the first in the response, however, that might not really work for all exchanges.
>     try getting the historical public trades for a market to determine the start of the trading (and, effectively, the start of OHLCV history), but, again, not all exchanges serve their entire history.
> 
> The final solution should probably include all of the above approaches.

These are great ideas!

It's already bad enough if they don't use UTC for the daily close/open, but have you ever encountered an exchange with a huge time drift (daylight saving etc. sounds crazy, but who knows)?

I'm asking because I'm worrying about how often this information needs to be updated...

Thanks again!

**Comment by hanyoonLDA** [comment]
Not sure if this helps, but a very quick and dirty "self-serve" API for candlestick data using CCXT and Flask.

```
import ccxt  # noqa: E402
from flask import Flask, jsonify

app = Flask(__name__)
exchanges = {}  # a placeholder for your instances

for id in ccxt.exchanges:
    try:
        exchange = getattr(ccxt, id)
        exchanges[id] = exchange()
    except Exception as err:
        print("Skipping: " + str(err))

@app.route('/ticker/<exchange>/<symbol1>/<symbol2>/<timeframe>/<sincetime>/<limit>', methods=['GET'])
def get_ohlcv(exchange,symbol1,symbol2, timeframe, sincetime, limit):
    if exchanges[exchange].has["fetchOHLCV"]:
        print(exchange + ' timeframes: ' + str(exchanges[exchange].timeframes))
        sincetime = int(sincetime) * 1000
        return jsonify(exchanges[exchange].fetch_ohlcv(symbol1+"/"+symbol2, timeframe, since=int(sincetime), limit=int(limit)))
    else:
        return 'Sorry, this exchange does not support OHLCV querying.'
# now exchanges dictionary contains all exchange instances...

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=80) #for open servers. Remove host and port if you plan on running on localhost.
```

This should help with quick data aggregation and seeing which exchanges support what.

**Comment by EkremSel** [comment]
Thanks for a great thread. I have resampled 1min to 4H by creating a dict for resample.agg() function.I am only getting the close price of the symbol. It works with resample.apply() also, However, when resampling into 4hr, the 4hr period periods does not match with the exchange. There is a +1 hour difference. I have tried using since=int(timestamp)*1000 to no avail. 

Below is my code: 

`data1 = exchange.fetch_ohlcv(symbol1,"1h",since=int(1664586000)*1000) `
`c_dic={'symbol1':'last','symbol2':'last'}`
`close=close.resample('4H').agg(c_dic)`

> 	Date Time		
2022-10-01 08:00:00	-0.00323741	0.00934944
2022-10-01 12:00:00	-0.00868469	0.00045756
2022-10-01 16:00:00	-0.01372266	-0.00183150
2022-10-01 20:00:00	-0.00851069	-0.01928031
2022-10-02 00:00:00	-0.01233891	0.00899231
...	...	...
2022-11-07 00:00:00	-0.06109711	-0.05123599
2022-11-07 04:00:00	0.00823145	0.01238538
2022-11-07 08:00:00	0.01111661	0.01975269
2022-11-07 12:00:00	0.03487430	0.00856359
2022-11-07 16:00:00	-0.00951681	-0.01009062

Instead, the tail should be :

> 07:00:00
11:00:00
15:00:00
